packages/next/src/server/after/after-context.ts

```diff
@@ -70,6 +71,16 @@ export class AfterContext {
       this.workUnitStores.add(workUnitStore)
     }
 
+    const afterTaskStore = afterTaskAsyncStorage.getStore()
+
+    // This is used for checking if request APIs can be called inside `after`.
+    // Note that we need to check the phase in which the *topmost* `after` was called (which should be "action"),
+    // not the current phase (which might be "after" if we're in a nested after).
+    // Otherwise, we might allow `after(() => headers())`, but not `after(() => after(() => headers()))`.
+    const rootTaskSpawnPhase = afterTaskStore
+      ? afterTaskStore.rootTaskSpawnPhase // nested after
+      : workUnitStore?.phase // topmost after
+
```

Comment:
We don't have great isolation already between action/render. If you put a non-blocking async function in an action and it is delayed enough that it doesn't run until the render phase has started you could get racey behavior where the speed of resolution determines the semantics of the after call. I think we decided this was ok b/c what useful thing could you do by spawning non-blockign async work in an action.

But I think you are right we need this new storage however unfortunate it might be though maybe it would be simpler to model it as a non-shadowing storage. Like conditionally run `bindSnapshot` inside a `afterTaskAsyncStorage` only at the top level when you are inside a store that is `render` phase. If there are any nested after calls they won't create deeper afterTask store instances because they're not in `render` phase so they'll naturally inherit the root phase

Reply:
> If you put a non-blocking async function in an action and it is delayed enough that it doesn't run until the render phase has started you could get racey behavior where the speed of resolution determines the semantics of the after call

sure yeah but i don't think we can do a lot about that, it's racey bc the user wrote racey code 🤷‍♀️ 

> maybe it would be simpler to model it as a non-shadowing storage. [...] If there are any nested after calls they won't create deeper afterTask store instances

soon enough i'm going to shadow it anyway to pass along callstack information (in dev)
https://github.com/vercel/next.js/pull/73449/files#diff-d6514c30f67dede0884866e69aa5c1ce48326f9fa2d515b5eb5e7162b5b0eb14R107
(i need this so that we can stitch together a sensible stack for the `after` callback -- instead of just showing `p-queue` at the top, tying it back to the place that initiate it)